pyface/data_view/i_data_view_widget.py

```diff
@@ -8,52 +8,283 @@
 #
 # Thanks for using Enthought open source!
 
-from traits.api import Bool, HasStrictTraits, Instance
+from contextlib import contextmanager
+import logging
+
+from traits.api import (
+    Bool, ComparisonMode, Enum, HasStrictTraits, Instance, List, Property,
+    TraitError, Tuple, cached_property,
+)
 
 from pyface.data_view.abstract_data_model import AbstractDataModel
 from pyface.i_widget import IWidget
 
 
+logger = logging.getLogger(__name__)
+logger.level = logging.DEBUG
+
+
 class IDataViewWidget(IWidget):
+    """ Interface for data view widgets. """
+
+    #: The data model for the data view.
+    data_model = Instance(AbstractDataModel, allow_none=False)
 
     #: Whether or not the column headers are visible.
     header_visible = Bool(True)
 
-    #: The data model for the data view.
-    data_model = Instance(AbstractDataModel, allow_none=False)
+    #: What can be selected.  Some backends may support more.
+    selection_type = Enum("row",)
+
+    #: How selections are modified.
+    selection_mode = Enum("extended", "single")
```

Comment:
Unfortunately Wx doesn't support "none" and a couple of other modes that may be useful.  Similarly with column and cell selection types.  I went through several iterations of how to handle this and ended up with putting the minimal set into the traits, but supporting more types in the implementation.

The other option is to give the full set of possibilities here rather than the minimal set, and have a more restrictive enum for the Wx implementation, but that breaks the Liskov substitution principle.

I will add something to the docstrings to clarify that subclasses may accept more.

Reply:
But this smaller set of values on the interface is quite crippling when it comes to writing tests for toolkit agnostic code. e.g. I could not test the validation logic in the mixin class unless this enum actually allows all the values expected here. Some branches in the validation method are therefore untested.
The wx implementation can handle the unsupported mode gracefully in its event handler, right?

I still think it would be a nicer API to say "these are all the values the implementation is supposed to handle, but they might decide to fall back to a default if a chosen mode cannot be supported."